dbos/system_database.go

```diff
@@ -905,6 +974,270 @@ func (s *systemDatabase) GetWorkflowSteps(ctx context.Context, workflowID string
+		output:        nil,
+		err:           nil,
+		tx:            tx,
+	}
+
+	err = s.RecordOperationResult(ctx, recordInput)
+	if err != nil {
+		return fmt.Errorf("failed to record operation result: %w", err)
+	}
+
+	// Commit transaction
+	if err := tx.Commit(ctx); err != nil {
+		return fmt.Errorf("failed to commit transaction: %w", err)
+	}
+
+	return nil
+}
+
+// Recv is a special type of step that receives a message destined for a given workflow
+func (s *systemDatabase) Recv(ctx context.Context, input WorkflowRecvInput) (any, error) {
+	functionName := "DBOS.recv"
+
+	// Get workflow state from context
+	// XXX these checks might be better suited for outside of the system db code. We'll see when we implement the client.
+	workflowState, ok := ctx.Value(WorkflowStateKey).(*WorkflowState)
+	if !ok || workflowState == nil {
+		return nil, NewStepExecutionError("", functionName, "workflow state not found in context: are you running this step within a workflow?")
+	}
+
+	if workflowState.isWithinStep {
+		return nil, NewStepExecutionError(workflowState.WorkflowID, functionName, "cannot call Recv within a step")
+	}
+
+	stepID := workflowState.NextStepID()
+	destinationID := workflowState.WorkflowID
+
+	// Set default topic if not provided
+	topic := DBOS_NULL_TOPIC
+	if len(input.Topic) > 0 {
+		topic = input.Topic
+	}
+
+	tx, err := s.pool.Begin(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to begin transaction: %w", err)
+	}
+	defer tx.Rollback(ctx)
+
+	// Check if operation was already executed
+	// XXX this might not need to be in the transaction
+	checkInput := CheckOperationExecutionDBInput{
+		workflowID:   destinationID,
+		operationID:  stepID,
+		functionName: functionName,
+		tx:           tx,
+	}
+	recordedResult, err := s.CheckOperationExecution(ctx, checkInput)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check operation execution: %w", err)
+	}
+	if recordedResult != nil {
+		if recordedResult.output != nil {
+			return recordedResult.output, nil
+		}
+		return nil, fmt.Errorf("no output recorded in the last recv")
+	}
+
+	// First check if there is already a message available in the database.
+	// If not, we'll wait for a notification and timeout
+	var exists bool
+	query := `SELECT EXISTS (SELECT 1 FROM dbos.notifications WHERE destination_uuid = $1 AND topic = $2)`
+	err = tx.QueryRow(ctx, query, destinationID, topic).Scan(&exists)
+	if err != nil {
+		return false, fmt.Errorf("failed to check message: %w", err)
+	}
+	if !exists {
+		// We'll do so through registering a channel with the notification listener loop.
+		payload := fmt.Sprintf("%s::%s", destinationID, topic)
+		c := make(chan bool, 1) // Make it buffered to allow the notification listener to post a signal even if the receiver has not reached its select statement yet
+		_, loaded := s.notificationsMap.LoadOrStore(payload, c)
+		if loaded {
+			close(c)
+			fmt.Println("Receive already called for workflow ", destinationID)
```

Comment:
I think this check is better happening before checking the notifications table (see Python https://github.com/dbos-inc/dbos-transact-py/blob/main/dbos/_sys_db.py#L1307)

The reason is to fail faster to avoid more load on the database.